QIR-README.md

```diff
@@ -0,0 +1,188 @@
+  becomes the name of the global LLVM function that implements this instruction.
+- The signature of the LLVM function is derived directly from the Q# operation's signature.
+- Target-level operations should only have `body` specializations, not `adjoint` or `controlled`.
+- The name of the target-level instructions can be arbitrary. By keeping it different from the
+  name of the user-level instruction, we remove the need to explicitly qualify the name with the
+  namespace when we refer to it in the user-level instruction definition.
+- At the moment, the intrinsic can only have a `body` specialization. This may change in the
+  future, but also might not.
+
+### User-level instruction set
+
+Generally, user-level instructions should be defined in the `Microsoft.Quantum.Intrinsic`
+namespace to match standard Q# usage.
+Similarly, the instruction names should match the names in the standard Q# library in that
+namespace, so that other existing Q# code will find the instructions correctly.
+
+User-level instruction definitions follow this pattern:
+
+```qsharp
+    @Inline()
+    operation X(qb : Qubit) : Unit
+    is Adj {
+        body
+        {
+            PhysX(qb);  
+        }
+        adjoint self;
+    }
+```
+
+- The `Inline` attribute is optional. If it appears, the QIR generator will in-line the
+  implementation of the user-level instruction, so that the generated QIR directly calls the
+  target-level instruction. Otherwise the generated QIR will call into the LLVM function that
+  implements the user-level instruction, which will contain a call into the target-level
+  instruction.
+- The implementation of the user-level instruction can contain arbitrary Q# code. For instance,
+  both sample files define the adjoint of a rotation as the rotation by the negative of the angle.
+  Once I add support for controlled, the controlled `X` implementation will have Q# code that will
+  start by checking the number of control qubits and either calling a target-level `CNOT` if 
+  there's one, executing one of the various Toffoli decompositions if there are two, and using
+  multiple Toffolis with an ancilla qubit if there are more than two.
+- User-level instructions will often have both `adjoint` and `controlled` specializations. Because
+  target-level instructions don't have such specializations, the only directive that can be used
+  to define these specializations is `self` to mark self-adjoint instructions.
+
+## QIR Specification
+
+The QIR specification is on the [Q# language repository](https://github.com/microsoft/qsharp-language/tree/main/Specifications/QIR).
+
+## To-Dos
+
+- [ ] Add debug information to QIR, issue #637.
+- [ ] Generate a `main` function for an entrypoint, issue #638
+- [ ] Handle structure/tuple mapping for entrypoints, issue #639
+- [ ] Add metadata to QIR from Q# attributes, issue #640
+- [ ] Code clean-up, issue #641
+- [ ] Port to Linux and MacOS, issue #642
```

Comment:
Consider adding a note about `Range` (by Cesar to one of the recent PRs).

Reply:
This list also needs to be updated - I have a much more current one.

Reply:
done.